File: src/Application/Contributions/ContributionMonthRange.cs
using System.Text.RegularExpressions;
using MySociety.Application.Common.Exceptions;

namespace MySociety.Application.Contributions;

public static partial class ContributionMonthRange
{
    public const int MaxMonths = 24;

    public static bool TryValidate(
        string fromMonth,
        string toMonth,
        out string periodKey,
        out int monthCount,
        out string error)
    {
        periodKey = string.Empty;
        monthCount = 0;

        if (!TryParseMonth(fromMonth, out var fromYear, out var fromMon))
        {
            error = "From month must be yyyy-MM (e.g. 2026-05).";
            return false;
        }

        if (!TryParseMonth(toMonth, out var toYear, out var toMon))
        {
            error = "To month must be yyyy-MM (e.g. 2026-07).";
            return false;
        }

        var fromIndex = fromYear * 12 + fromMon;
        var toIndex = toYear * 12 + toMon;
        if (fromIndex > toIndex)
        {
            error = "From month cannot be after to month.";
            return false;
        }

        monthCount = toIndex - fromIndex + 1;
        if (monthCount > MaxMonths)
        {
            error = $"A contribution request can cover at most {MaxMonths} months.";
            return false;
        }

        periodKey = BuildPeriodKey(fromMonth.Trim(), toMonth.Trim());
        error = string.Empty;
        return true;
    }

    public static string BuildPeriodKey(string fromMonth, string toMonth) =>
        $"{fromMonth.Trim()}..{toMonth.Trim()}";

    public static int CountInclusiveMonths(string fromMonth, string toMonth)
    {
        if (!TryParseMonth(fromMonth, out var fromYear, out var fromMon) ||
            !TryParseMonth(toMonth, out var toYear, out var toMon))
        {
            throw new ValidationException("Invalid month range.");
        }

        return toYear * 12 + toMon - (fromYear * 12 + fromMon) + 1;
    }

    public static string AddMonths(string yyyyMm, int monthsToAdd)
    {
        if (!TryParseMonth(yyyyMm, out var year, out var month))
        {
            throw new ValidationException("Invalid month.");
        }

        var index = year * 12 + (month - 1) + monthsToAdd;
        var newYear = index / 12;
        var newMonth = index % 12 + 1;
        return $"{newYear:D4}-{newMonth:D2}";
    }

    public static IReadOnlyList<string> GetRecentMonths(int count = 12, DateTime? utcNow = null)
    {
        var now = utcNow ?? DateTime.UtcNow;
        var months = new List<string>(count);
        var year = now.Year;
        var month = now.Month;

        for (var i = 0; i < count; i++)
        {
            months.Add($"{year:D4}-{month:D2}");
            if (month == 1)
            {
                year--;
                month = 12;
            }
            else
            {
                month--;
            }
        }

        months.Reverse();
        return months;
    }

    public static string FormatDisplayLabel(string period)
    {
        if (period.Contains("..", StringComparison.Ordinal))
        {
            var parts = period.Split("..", 2, StringSplitOptions.TrimEntries);
            if (parts.Length == 2 &&
                TryParseMonth(parts[0], out var fy, out var fm) &&
                TryParseMonth(parts[1], out var ty, out var tm))
            {
                var count = CountInclusiveMonths(parts[0], parts[1]);
                return $"{FormatMonth(fy, fm)} – {FormatMonth(ty, tm)} ({count} mo)";
            }
        }

        if (TryParseMonth(period, out var y, out var m))
        {
            return FormatMonth(y, m);
        }

        return period;
    }

    private static string FormatMonth(int year, int month) =>
        new DateTime(year, month, 1).ToString("MMM yyyy");

    public static bool TryParseMonth(string value, out int year, out int month)
    {
        year = 0;
        month = 0;
        value = value.Trim();
        var match = MonthRegex().Match(value);
        if (!match.Success)
        {
            return false;
        }

        year = int.Parse(match.Groups[1].Value);
        month = int.Parse(match.Groups[2].Value);
        return true;
    }

    public static bool TryParsePeriodKey(string periodKey, out int fromIndex, out int toIndex)
    {
        fromIndex = 0;
        toIndex = 0;

        var trimmed = periodKey.Trim();
        string fromMonth;
        string toMonth;

        if (trimmed.Contains("..", StringComparison.Ordinal))
        {
            var parts = trimmed.Split("..", 2, StringSplitOptions.TrimEntries);
            if (parts.Length != 2)
            {
                return false;
            }

            fromMonth = parts[0];
            toMonth = parts[1];
        }
        else
        {
            fromMonth = trimmed;
            toMonth = trimmed;
        }

        if (!TryParseMonth(fromMonth, out var fromYear, out var fromMon) ||
            !TryParseMonth(toMonth, out var toYear, out var toMon))
        {
            return false;
        }

        fromIndex = ToMonthIndex(fromYear, fromMon);
        toIndex = ToMonthIndex(toYear, toMon);
        return fromIndex <= toIndex;
    }

    public static bool MonthRangesOverlap(string fromMonth, string toMonth, string existingPeriodKey)
    {
        if (!TryParseMonth(fromMonth, out var fromYear, out var fromMon) ||
            !TryParseMonth(toMonth, out var toYear, out var toMon) ||
            !TryParsePeriodKey(existingPeriodKey, out var existingFromIndex, out var existingToIndex))
        {
            return false;
        }

        var requestFromIndex = ToMonthIndex(fromYear, fromMon);
        var requestToIndex = ToMonthIndex(toYear, toMon);
        return requestFromIndex <= existingToIndex && existingFromIndex <= requestToIndex;
    }

    public static bool TryFindOverlappingPeriod(
        string fromMonth,
        string toMonth,
        IEnumerable<string> existingPeriodKeys,
        out string overlappingPeriod)
    {
        foreach (var existing in existingPeriodKeys)
        {
            if (MonthRangesOverlap(fromMonth, toMonth, existing))
            {
                overlappingPeriod = existing;
                return true;
            }
        }

        overlappingPeriod = string.Empty;
        return false;
    }

    private static int ToMonthIndex(int year, int month) => year * 12 + month;

    [GeneratedRegex(@"^(\d{4})-(0[1-9]|1[0-2])$")]
    private static partial Regex MonthRegex();
}

File: mobile/src/utils/contributionMonthRange.ts
const MONTH_REGEX = /^(\d{4})-(0[1-9]|1[0-2])$/;

export function getRecentMonths(count = 12, date = new Date()): string[] {
  const months: string[] = [];
  let year = date.getFullYear();
  let month = date.getMonth() + 1;

  for (let i = 0; i < count; i++) {
    months.push(`${year}-${String(month).padStart(2, '0')}`);
    if (month === 1) {
      year -= 1;
      month = 12;
    } else {
      month -= 1;
    }
  }

  return months.reverse();
}

export function getCurrentMonth(date = new Date()): string {
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;
}

export function addMonths(yyyyMm: string, monthsToAdd: number): string {
  const parsed = parseMonth(yyyyMm);
  if (!parsed) throw new Error('Invalid month');
  const index = parsed.year * 12 + (parsed.month - 1) + monthsToAdd;
  const year = Math.floor(index / 12);
  const month = (index % 12) + 1;
  return `${year}-${String(month).padStart(2, '0')}`;
}

export function countInclusiveMonths(fromMonth: string, toMonth: string): number {
  const from = parseMonth(fromMonth);
  const to = parseMonth(toMonth);
  if (!from || !to) return 0;
  return to.year * 12 + to.month - (from.year * 12 + from.month) + 1;
}

export function buildPeriodKey(fromMonth: string, toMonth: string): string {
  return `${fromMonth}..${toMonth}`;
}

export function formatContributionPeriodLabel(period: string): string {
  if (period.includes('..')) {
    const [from, to] = period.split('..');
    if (from && to) {
      const count = countInclusiveMonths(from, to);
      return `${formatSingleMonth(from)} – ${formatSingleMonth(to)} (${count} mo)`;
    }
  }

  return formatSingleMonth(period) || period;
}

function formatSingleMonth(yyyyMm: string): string {
  const parsed = parseMonth(yyyyMm);
  if (!parsed) return yyyyMm;
  return new Date(parsed.year, parsed.month - 1, 1).toLocaleDateString('en-IN', {
    month: 'short',
    year: 'numeric',
  });
}

function parseMonth(value: string): { year: number; month: number } | null {
  const match = value.trim().match(MONTH_REGEX);
  if (!match) return null;
  return { year: Number(match[1]), month: Number(match[2]) };
}

export function parseMonthKey(value: string): { year: number; month: number } | null {
  return parseMonth(value);
}

export function formatMonthKey(year: number, month: number): string {
  return `${year}-${String(month).padStart(2, '0')}`;
}

export const MONTH_PICKER_OPTIONS = Array.from({ length: 12 }, (_, index) => {
  const month = index + 1;
  return {
    value: String(month).padStart(2, '0'),
    label: new Date(2000, index, 1).toLocaleDateString('en-IN', { month: 'long' }),
  };
});

export function buildYearOptions(minYear: number, maxYear: number): number[] {
  const years: number[] = [];
  for (let year = minYear; year <= maxYear; year += 1) {
    years.push(year);
  }
  return years;
}

export function getContributionYearRange(date = new Date()): { minYear: number; maxYear: number } {
  const year = date.getFullYear();
  return { minYear: year - 10, maxYear: year + 10 };
}

function toMonthIndex(year: number, month: number): number {
  return year * 12 + month;
}

export function parsePeriodKey(periodKey: string): { fromIndex: number; toIndex: number } | null {
  const trimmed = periodKey.trim();
  let fromMonth: string;
  let toMonth: string;

  if (trimmed.includes('..')) {
    const parts = trimmed.split('..');
    if (parts.length !== 2 || !parts[0] || !parts[1]) {
      return null;
    }
    fromMonth = parts[0];
    toMonth = parts[1];
  } else {
    fromMonth = trimmed;
    toMonth = trimmed;
  }

  const from = parseMonth(fromMonth);
  const to = parseMonth(toMonth);
  if (!from || !to) {
    return null;
  }

  const fromIndex = toMonthIndex(from.year, from.month);
  const toIndex = toMonthIndex(to.year, to.month);
  if (fromIndex > toIndex) {
    return null;
  }

  return { fromIndex, toIndex };
}

export function monthRangesOverlap(
  fromMonth: string,
  toMonth: string,
  existingPeriodKey: string,
): boolean {
  const from = parseMonth(fromMonth);
  const to = parseMonth(toMonth);
  const existing = parsePeriodKey(existingPeriodKey);
  if (!from || !to || !existing) {
    return false;
  }

  const requestFromIndex = toMonthIndex(from.year, from.month);
  const requestToIndex = toMonthIndex(to.year, to.month);
  return requestFromIndex <= existing.toIndex && existing.fromIndex <= requestToIndex;
}

export function findOverlappingPeriod(
  fromMonth: string,
  toMonth: string,
  existingPeriodKeys: string[],
): string | null {
  for (const existing of existingPeriodKeys) {
    if (monthRangesOverlap(fromMonth, toMonth, existing)) {
      return existing;
    }
  }
  return null;
}

export const DURATION_OPTIONS = [1, 2, 3, 4, 6, 9, 12] as const;
